capsules/src/pwm.rs

```diff
@@ -0,0 +1,137 @@
+use kernel::grant::{AllowRoCount, AllowRwCount, Grant, UpcallCount};
+use kernel::hil;
+use kernel::syscall::{CommandReturn, SyscallDriver};
+use kernel::utilities::cells::OptionalCell;
+use kernel::{ErrorCode, ProcessId};
+
+/// Syscall driver number.
+use crate::driver;
+pub const DRIVER_NUM: usize = driver::NUM::Pwm as usize;
+
+// An empty app, for potential uses in future updates of the driver
+#[derive(Default)]
+pub struct App;
+
+pub struct Pwm<'a, const NUM_PINS: usize> {
+    /// The usable pwm pins.
+    pwm_pins: &'a [&'a dyn hil::pwm::PwmPin; NUM_PINS],
+    /// Per-app state.
+    apps: Grant<App, UpcallCount<1>, AllowRoCount<0>, AllowRwCount<0>>,
+    /// An array of apps associated to their reserved pins.
+    active_app: [OptionalCell<ProcessId>; NUM_PINS],
+}
+
+impl<'a, const NUM_PINS: usize> Pwm<'a, NUM_PINS> {
+    pub fn new(
+        pwm_pins: &'a [&'a dyn hil::pwm::PwmPin; NUM_PINS],
+        grant: Grant<App, UpcallCount<1>, AllowRoCount<0>, AllowRwCount<0>>,
+    ) -> Pwm<'a, NUM_PINS> {
+        assert!(NUM_PINS <= (u16::MAX as usize));
+        const EMPTY: OptionalCell<ProcessId> = OptionalCell::empty();
+        Pwm {
+            pwm_pins: pwm_pins,
+            apps: grant,
+            active_app: [EMPTY; NUM_PINS],
+        }
+    }
+
+    pub fn is_valid_app(&self, processid: ProcessId, pin: usize) -> bool {
+        // Attempt to get the app that is using the pin.
+        self.active_app[pin].map_or(true, |id| {
+            // If the app is empty, that means that there is no app currently using the pins,
+            // therefore the pin is usable by the new app
+            if id == &processid {
+                // The same app is trying to access the pin it has access to, valid
+                true
+            } else {
+                // An app is trying to access another app's pin, invalid
+                false
+            }
+        })
+    }
+}
+
+/// Provide an interface for userland.
+impl<'a, const NUM_PINS: usize> SyscallDriver for Pwm<'a, NUM_PINS> {
+    /// Command interface.
+    ///
+    /// ### `command_num`
+    ///
+    /// - `0`: Return number of PWM pins if this driver is included on the platform.
+    /// - `1`: Start the PWM pin output. First 16 bits of `data1` are used for the duty cycle, as a
+    ///     percentage with 2 decimals, and the last 16 bits of `data2` are used for the PWM channel
+    ///     to be controlled. `data2` is used for the frequency in hertz. For the duty cycle, 100% is
+    ///     the max duty cycle for this pin.
+    /// - `2`: Stop the PWM output.
+    /// - `3`: Return the maximum possible frequency for this pin.
+    fn command(
+        &self,
+        command_num: usize,
+        data1: usize,
+        data2: usize,
+        processid: ProcessId,
+    ) -> CommandReturn {
+        match command_num {
+            // Check whether the driver exists, return number of usable PWM pins.
+            0 => CommandReturn::success_u32(NUM_PINS as u32),
```

Comment:
Please separate out these two pieces of functionality. 0 should be just an "exists" call, there should be a separate command for how many pins.

Reply:
I kind of disagree with this, as:
- ADC https://github.com/tock/tock/blob/500caa5e882b7efe40158849fb489205f07c6a2e/capsules/src/adc.rs#L1149
- GPIO https://github.com/tock/tock/blob/500caa5e882b7efe40158849fb489205f07c6a2e/capsules/src/gpio.rs#L198
- LED https://github.com/tock/tock/blob/500caa5e882b7efe40158849fb489205f07c6a2e/capsules/src/led.rs#L95

are implemented in the same way.

It is true that the TRD104 specifies that [command 0](https://github.com/tock/tock/blob/master/doc/reference/trd104-syscalls.md#431-command-identifier-0) should always return `Success`, but this might not be clear if it is the success class of return values or the actual success without payload.

While I do not favor one or the other, we should keep implementations consistent for all drivers.

Reply:
Ah. Hunh. If this is consistent with some major system call capsules, then I am OK with it. My concern was about consistency, not that one way or the other is better.

Reply:
@TeodoraMiu please revert this so that command 0 returns the number of PWM pins

Reply:
It seems like all of these capsules are potentially non-compliant with trd104: https://github.com/tock/tock/blob/master/doc/reference/trd104-syscalls.md#431-command-identifier-0 . I vaguely recall some discussion of this back around Tock 2.0, and I think there was a desire for userspace to be able to assume the same system call return variant for all driver-exists checks

Reply:
cc @lschuermann I think you may have been a part of those discussions?